# STAT 207 Group Lab Assignment 8 - [10 total points]

## Confidence Interval Properties

<hr>

## <u>Lab Grading</u>:

Should we grade your submission?  If not, write the netID of the submission to be graded.  (Note: We will only grade one assignment per group, and we'll pick the first one that says we should grade that submission.  We will assign the same grade to all team members.)

*For example*, you might respond: **grade this submission** or **my submission is under netID jdeeke**

grade this submission

If you said **my submission is under netID** above, we will not read any more of your lab submission.

If you said **grade this submission** above, who worked with you on this submission?  Write both their **names** and **netIDs**.  

Also, discuss and record your favorite Pokemon, for those who play Pokemon.

Joonwoo Ha (ha47)

<hr>

## <u>Purpose</u>:
You should work in groups of 2-3 on this report (not working in groups without permission will result in a point deduction). The purpose of this group lab assignment is to explore properties of confidence intervals.
<hr>

## <u>Assignment Instructions</u>:


### Group Roles

Suggested and specified roles are provided below: 

#### Groups of 2

* **Driver**: This student will type the report.  While typing the report, you may be the one who is selecting the functions to apply to the data.
* **Navigator**: This student will guide the process of answering the question.  Specific ways to help may include: outlining the general steps needed to solve a question (providing the overview), locating examples within the course notes, and reviewing each line of code as it is typed.

#### Groups of 3

* **Driver**: This student will type the report.  They may also be the one to select the functions to apply to the data.
* **Navigator**: This student will guide the process of answering the question.  They may select the general approach to answering the question and/or a few steps to be completed along the way. 
* **Communicator**: This student will review the report (as it is typed) to ensure that it is clear and concise.  This student may also locate relevant examples within the course notes that may help complete the assignment.

<hr>

## 0. [0.5 pts] Imports

In [1]:
import numpy as np
import pandas as pd

<hr>

## <u>Case Study</u>: Pokemon Confidence Intervals

We previously explored the Pokemon dataset in Lab 5.  Recall that this dataset contains information about all current Pokemon, including those recently released in Generation 9 (2023).  

The variables in the data include:

- **Height**: height (m)
- **Weight**: weight (kg)
- **Type1**: the primary type of the Pokemon
- **Ability1**: the primary ability of the Pokemon
- **Ability_Hidden**: the hidden ability of the Pokemon, if one is present
- **Get_Rate**: the rate at which the Pokemon can be captured (higher = easier to capture)
- **HP**: the hit point, which is how much damage a Pokemon can sustain before fainting
- **Attack**: the damage a Pokemon can inflict with a physical move
- **Defense**: the damage a Pokemon receives when hit with a physical move
- **SP_Attack**: the damage a Pokemon can inflict with a special move
- **SP_Defense**: the damage a Pokemon receives when hit with a special move
- **Speed**: speed, which determines the order in which Pokemon can move
- **Total**: the sum of each of the base statistics (HP, Attack, Defense, SP_Attack, SP_Defense, and Speed).

There are other variables in the data, which you are allowed to use if you understand their provided values.  More information about this dataset can be found at https://www.kaggle.com/datasets/takamasakato/pokemon-all-status-data.

We'll also use a second dataset, `deeke.csv`, which has information about the 104 Pokemon that my family has collected.

## 1. [1.5 pt] Population Information

**a)** Read in the population data.

In [7]:
df = pd.read_csv("Pokedex_Ver9.csv")
df2 = pd.read_csv("deeke.csv")
df

No Branch_Code Original_Name           Name  Generation  Height  \
0        1         1_0     Bulbasaur      Bulbasaur           1     0.7   
1        2         2_0       Ivysaur        Ivysaur           1     1.0   
2        3         3_0      Venusaur       Venusaur           1     2.0   
3        3         3_1      Venusaur  Mega Venusaur           1     2.4   
4        4         4_0    Charmander     Charmander           1     0.6   
...    ...         ...           ...            ...         ...     ...   
1198  1006      1006_0  Iron Valiant   Iron Valiant           9     1.4   
1199  1007      1007_0      Koraidon       Koraidon           9     2.5   
1200  1008      1008_0      Miraidon       Miraidon           9     3.5   
1201  1009      1009_0  Walking Wake   Walking Wake           9     3.5   
1202  1010      1010_0   Iron Leaves    Iron Leaves           9     1.5   

      Weight     Type1     Type2          Ability1  ... SP_Attack SP_Defense  \
0        6.9     Grass    Poison          Overgrow  ...        65         65   
1       13.0     Grass    Poison          Overgrow  ...        80         80   
2      100.0     Grass    Poison          Overgrow  ...       100        100   
3      155.5     Grass    Poison         Thick Fat  ...       122        120   
4        8.5      Fire       NaN             Blaze  ...        60         50   
...      ...       ...       ...               ...  ...       ...        ...   
1198    35.0     Fairy  Fighting       Quark Drive  ...       120         60   
1199   303.0  Fighting    Dragon  Orichalcum Pulse  ...        85        100   
1200   240.0  Electric    Dragon     Hadron Engine  ...       135        115   
1201   280.0     Water    Dragon    Protosynthesis  ...       125         83   
1202   125.0     Grass   Psychic       Quark Drive  ...        70        108   

     Speed  Total  E_HP  E_Attack  E_Defense E_SP_Attack E_SP_Defense  E_Speed  
0       45    318     0         0          0           1            0        0  
1       60    405     0         0          0           1            1        0  
2       80    525     0         0          0           2            1        0  
3       80    625     0         0          0           2            1        0  
4       65    309     0         0          0           0            0        1  
...    ...    ...   ...       ...        ...         ...          ...      ...  
1198   116    590     0         3          0           0            0        0  
1199   135    670     0         3          0           0            0        0  
1200   135    670     0         0          0           3            0        0  
1201   109    590     0         0          0           3            0        0  
1202   104    590     0         3          0           0            0        0  

[1203 rows x 38 columns]

**b)** Pick a variable and a reasonable characteristic for that variable to explore for this lab.  

For the population of all Pokemon, calculate your characteristic for your variable of interest.  This is your parameter of interest for this lab.


In [6]:
df['Height'].describe()

count    1203.000000
mean        1.370158
std         3.163398
min         0.100000
25%         0.600000
50%         1.000000
75%         1.600000
max       100.000000
Name: Height, dtype: float64

**c)** Define your parameter of interest.  That is, report the value and what it means in a sentence.

The parameter of interest in this analysis is the mean height of all Pokemon, 1.37 meters. This value represents the average height of Pokemon in the dataset. The standard deviation of 3.16 shows  that there is a high variance in height, meaning that some Pokemon are significantly taller or shorter than the average. The median height of 1.0 meters means that half of the Pokemon in the dataset are shorter than or equal to 1.0 meters, while the other half are taller.

## 2. [4 pts] Creating a Confidence Interval

While we technically have the population of all Pokemon (and therefore can-and have-calculated our true parameter of interest), we will pretend that you only have the information available to my family about Pokemon.  To do so, you will use my family's Pokemon data to generate a confidence interval for your parameter of interest that you selected in **1b**.  By having the population, you can then "check" our confidence interval to see how well it performs. 

**a)** Provide your best estimate for the population parameter based on my family's data.

In [ ]:
estimate = df2["Height"].mean()
print("Best estimate for population parameter:", estimate)

Best estimate for population parameter: 0.9971153846153846


**b)** Using my sample of size 104, simulate an estimated sampling distribution for your characteristic of interest from **1b**.

In [9]:
sample_size = 104
num_simulations = 1000
sample_means = [df2["Height"].sample(sample_size, replace=True).mean() for _ in range(num_simulations)]

print("Mean of sampling distribution:", np.mean(sample_means))
print("Standard deviation of sampling distribution:", np.std(sample_means))


Mean of sampling distribution: 0.9939201923076924
Standard deviation of sampling distribution: 0.07556624674419418


**c)** From your simulated sampling distribution, calculate an 85% confidence interval for your parameter of interest.

In [10]:
lower_bound = np.percentile(sample_means, 7.5)
upper_bound = np.percentile(sample_means, 92.5)

print("85% Confidence Interval:", (lower_bound, upper_bound))


85% Confidence Interval: (0.8894230769230769, 1.1068028846153846)


**d)** For the first time in this assignment, we will recognize that we do have access to our population data and therefore can access the parameter of interest.

Is your confidence interval "good"?  That is, does your confidence interval capture the value for the true parameter of interest?

**For full credit, find the answer to this question using code & logical statements in one line of code**.

In [11]:
print("Is the confidence interval good?", lower_bound <= df["Height"].mean() <= upper_bound)

Is the confidence interval good? False


The fact that the answer to this code is false tells us that the confidence interval does not include the population mean. It is possible that the data we used (deeke.csv) are not well represented in the population.

## 3. [3 pt] Creating Repeated Confidence Intervals

In Question 2, we generated a confidence interval based on my family's data.  My family has captured 104 Pokemon, but what about other Pokemon players who have also captured 104 Pokemon?  What would their intervals look like?

We'd like to learn more about the behavior of confidence intervals in general, so we'd like to observe what happens when we have different samples of Pokemon throughout this question.

For this question, do the following:

- Generate 1000 random samples (each of size 104 without replacement) from the population of all Pokemon
- Calculate the 85% confidence interval for each sample
- Record whether the confidence interval contains the population parameter

Be sure to record whether the confidence interval contains the population parameter.  You may also record other information, although this is not required.

*Hint 1: Using two `for` loops, where one occurs inside the other will likely help you with this process.*

*Hint 2: As you approach this process, you may want to look back on the code you used in <b>Q2b - 2d</b>*.

*Hint 3: If you find that you are stuck, try first to complete the steps described above for a single random sample.  Then, extend the process 1000 times using a `for` loop.*

In [15]:
true_mean = df["Height"].mean()

num_samples = 1000
sample_size = 104
num_bootstrap = 1000

count_true_mean = 0

confidence_intervals = []

for i in range(num_samples):
    sample = df["Height"].sample(sample_size, replace=False)
    sample_means = []
    for j in range(num_bootstrap):
        resample = sample.sample(sample_size, replace=True)
        sample_means.append(resample.mean())

    lower_bound = np.percentile(sample_means, 7.5)
    upper_bound = np.percentile(sample_means, 92.5)

    contains_true_mean = lower_bound <= true_mean <= upper_bound
    if contains_true_mean:
        count_true_mean += 1
    confidence_intervals.append((lower_bound, upper_bound, contains_true_mean))

print(f"Out of {num_samples} samples, {count_true_mean} confidence intervals contained the true mean.")


Out of 1000 samples, 732 confidence intervals contained the true mean.


## 4. [1 pt] Calculating the Proportion of Confidence Intervals Containing Our Parameter

**a)** From the output of the 1000 random samples, calculate the proportion of confidence intervals that contain our parameter.

In [16]:
proportion_true_mean = count_true_mean / num_samples

print("Proportion of confidence intervals containing the true mean:", proportion_true_mean)


Proportion of confidence intervals containing the true mean: 0.732


**b)** Does the proportion that you calculated in part **a** match with the anticipated behavior of confidence intervals?  Briefly explain.

The calculated proportion of confidence intervals that contain the true mean is 73.2%, which is lower than the expected 85%. This suggests that our confidence intervals are not having the population parameter as often as anticipated. Some variation is expected due to random sampling, but the lower proportion may indicate sampling variability or narrower confidence intervals.